### PR TITLE
Add Sign-in prompt

### PR DESCRIPTION
**Pull request by luizgrp**
#### WHAT

Add Sign-in prompt + sample.

#### Checklist :clipboard:
- [x] Add explicit visibility modifier and explicit return types for public declarations
- [x] Run spotless check
- [x] Run tests
- [x] Update metalava's signature text files

### CODE REVIEW

File: datalayer/sample/phone/src/main/java/com/google/android/horologist/datalayer/sample/screens/main/MainScreen.kt [hunk 1]
```diff
@@ -74,7 +74,10 @@ fun MainScreen(
                     InstallAppPromptDemo2Screen()
                 }
                 composable(route = Screen.ReEngagePromptDemoScreen.route) {
-                    ReEngagePromptDemoScreen()
+                    SignInPromptDemoScreen()
```
Comment (yschimke) [inline datalayer/sample/phone/src/main/java/com/google/android/horologist/datalayer/sample/screens/main/MainScreen.kt:77]:
```suggestion
                    ReEngagePromptDemoScreen()
```

File: datalayer/sample/phone/src/main/java/com/google/android/horologist/datalayer/sample/screens/inappprompts/signin/SignInPromptDemoViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        val signInPrompt: SignInPrompt,
+    ) : ViewModel() {
+
+        private var initializeCalled = false
+
+        private val _uiState =
+            MutableStateFlow<SignInPromptDemoScreenState>(SignInPromptDemoScreenState.Idle)
+        public val uiState: StateFlow<SignInPromptDemoScreenState> = _uiState
+
+        @MainThread
```
Comment (yschimke) [inline datalayer/sample/phone/src/main/java/com/google/android/horologist/datalayer/sample/screens/inappprompts/signin/SignInPromptDemoViewModel.kt:45]: Curious? why MainThread on these? Avoiding locking on initializeCalled?

If so, is it worth a comment on initializeCalled, "always accessed on Main thread"

File: datalayer/phone-ui/src/main/java/com/google/android/horologist/datalayer/phone/ui/prompt/signin/CoroutineScopeHolder.kt [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.horologist.datalayer.phone.ui.prompt.signin
+
+import kotlinx.coroutines.CoroutineScope
+
+internal object CoroutineScopeHolder {
```
Comment (yschimke) [inline datalayer/phone-ui/src/main/java/com/google/android/horologist/datalayer/phone/ui/prompt/signin/CoroutineScopeHolder.kt:21]: I'm still curious what this looks like in other similar libraries, but will take a look myself.

Not an issue in practice, or principle, I'm just curious about it.

**Review by yschimke — APPROVED** [review]

File: datalayer/sample/phone/src/main/java/com/google/android/horologist/datalayer/sample/screens/inappprompts/signin/SignInPromptDemoViewModel.kt (hunk 2 above)
Comment (luizgrp) [reply datalayer/sample/phone/src/main/java/com/google/android/horologist/datalayer/sample/screens/inappprompts/signin/SignInPromptDemoViewModel.kt:45]: Wouldn't the annotation be enough to indicate that? From the MainThread annotation kdocs:

> Denotes that the annotated method should only be called on the main thread. If the annotated element is a class, then all methods in the class should be called on the main thread.

Are you concerned the variable could be modified, internally within the class, by a different thread? I'd say if the variable is changed outside this method, it is already a red flag.

In the sample code from [DAC](https://developer.android.com/topic/architecture/ui-layer/state-production#initializing-state-production) there is a comment:
> This function is idempotent provided it is only called from the UI thread.

Let me know if you think we should add that.

File: datalayer/phone-ui/src/main/java/com/google/android/horologist/datalayer/phone/ui/prompt/signin/CoroutineScopeHolder.kt (hunk 3 above)
Comment (luizgrp) [reply datalayer/phone-ui/src/main/java/com/google/android/horologist/datalayer/phone/ui/prompt/signin/CoroutineScopeHolder.kt:21]: please let me know your findings 👍

File: datalayer/sample/phone/src/main/java/com/google/android/horologist/datalayer/sample/screens/main/MainScreen.kt (hunk 1 above)
Comment (luizgrp) [reply datalayer/sample/phone/src/main/java/com/google/android/horologist/datalayer/sample/screens/main/MainScreen.kt]: thanks!

File: datalayer/sample/phone/src/main/java/com/google/android/horologist/datalayer/sample/screens/inappprompts/signin/SignInPromptDemoViewModel.kt (hunk 2 above)
Comment (luizgrp) [reply datalayer/sample/phone/src/main/java/com/google/android/horologist/datalayer/sample/screens/inappprompts/signin/SignInPromptDemoViewModel.kt:45]: Merging but will come back to these open comments in the next iterations.

Comment (yschimke) [reply datalayer/sample/phone/src/main/java/com/google/android/horologist/datalayer/sample/screens/inappprompts/signin/SignInPromptDemoViewModel.kt:45]: Not an issue. More as part of the review, I was curious why it was MainThread, then reading the code realised it was the unsynchronized field.

But your call, nothing needed.

**Pull request merged**